components/brave_rewards/browser/rewards_p3a.cc

```diff
@@ -182,61 +182,57 @@ double CalcWalletBalance(base::flat_map<std::string, double> wallets,
-      // Probably they've been on since we started measuring.
-      // Ignore this interval since we can't measure it and treat it
-      // the same as never having enabled.
-      enabled_duration = RewardsEnabledDuration::kNever;
-    } else {
-      // Rewards have been enabled.
+    if (ads_enabled) {
+      // Ads have been enabled.
       // Remember when so we can measure the duration on later changes.
-      prefs->SetTime(prefs::kEnabledTimestamp, base::Time::Now());
-      enabled_duration = RewardsEnabledDuration::kStillEnabled;
+      prefs->SetTime(prefs::kAdsEnabledTimestamp, base::Time::Now());
+      enabled_duration = AdsEnabledDuration::kStillEnabled;
     }
   } else {
     // Previous timestamp available.
-    if (!rewards_enabled) {
-      // Rewards have been disabled. Record the duration they were on.
-      // Set the threshold at three units so each bin represents the
-      // nominal value as an order-of-magnitude: more than three days
-      // is a week, more than three weeks is a month, and so on.
-      constexpr int threshold = 3;
-      constexpr int days_per_week = 7;
-      constexpr double days_per_month = 30.44;  // average length
-      base::TimeDelta duration = base::Time::Now() - enabled_timestamp;
-      VLOG(1) << "Rewards disabled after " << duration;
-      if (duration < base::Hours(threshold)) {
-        enabled_duration = RewardsEnabledDuration::kHours;
-      } else if (duration < base::Days(threshold)) {
-        enabled_duration = RewardsEnabledDuration::kDays;
-      } else if (duration < base::Days(threshold * days_per_week)) {
-        enabled_duration = RewardsEnabledDuration::kWeeks;
-      } else if (duration < base::Days(threshold * days_per_month)) {
-        enabled_duration = RewardsEnabledDuration::kMonths;
-      } else {
-        enabled_duration = RewardsEnabledDuration::kQuarters;
-      }
+    if (!ads_enabled) {
+      // Ads have been disabled. Record the duration they were on.
+      enabled_time_delta = base::Time::Now() - enabled_timestamp;
+      VLOG(1) << "Rewards disabled after " << enabled_time_delta;
       // Null the timestamp so we're ready for a fresh measurement.
-      prefs->SetTime(prefs::kEnabledTimestamp, base::Time());
-    } else {
-      // Rewards have been enabled. Overwrite the previous timestamp.
-      // Normally we null the timestamp when rewards are disabled,
-      // so typically we'll take the other `else` branch above instead.
-      // We nevertheless mark a new timestamp here to maintain consistent
-      // measurement even if our prefs change through some other path.
-      prefs->SetTime(prefs::kEnabledTimestamp, base::Time::Now());
-      enabled_duration = RewardsEnabledDuration::kStillEnabled;
+      // Store the enabled time delta so we can keep reporting the duration.
+      prefs->SetTime(prefs::kAdsEnabledTimestamp, base::Time());
+      prefs->SetTimeDelta(prefs::kAdsEnabledTimeDelta, enabled_time_delta);
```

Comment:
It looks like saving this reports the last measured duration indefinitely, while P3A measurements are supposed to use information only from the last week. Maybe `kNever` should be `kNoDisable` or `kNoDuration` instead?

Reply:
The reason why I chose to store the duration, is because this function gets called whenever the browser is opened. Before these changes, the histogram value would get wiped out when this function is called at least twice after the user disables ads. The first call would result in the duration being reported correctly, and the `enabled_timestamp` being set to null. The second call would replace the histogram value with `kNever` because the `enabled_timestamp` is null, and no duration can be calculated (which is undesired).

The P3A service does not clear histogram values after its weekly report, and thus the same histogram values are reported forever (unless changed). This applies to all P3A metrics

Reply:
I see. Thanks for explaining my bug. :) Good catch!